_studio/mfx_lib/decode/mjpeg/src/mfx_mjpeg_dec_decode.cpp

```diff
@@ -2074,5 +2033,239 @@ mfxStatus VideoDECODEMJPEGBase_HW::CompleteTask(void *pParam, mfxStatus )
     return MFX_ERR_NONE;
 }
 
+#ifdef MFX_ENABLE_SW_FALLBACK
+VideoDECODEMJPEGBase_SW::VideoDECODEMJPEGBase_SW()
+{
+    m_FrameAllocator.reset(new mfx_UMC_FrameAllocator);
+    pLastTask = NULL;
+    m_tasksCount = 0;
+}
+
+mfxStatus VideoDECODEMJPEGBase_SW::Init(mfxVideoParam *decPar, mfxFrameAllocRequest *request, mfxFrameAllocResponse *response, mfxFrameAllocRequest *, bool isUseExternalFrames, VideoCORE *core)
+{
+    UMC::Status umcSts = m_FrameAllocator->InitMfx(0, core, decPar, request, response, isUseExternalFrames, true);
+    MFX_CHECK(umcSts == UMC::UMC_OK, MFX_ERR_MEMORY_ALLOC);
+
+    ConvertMFXParamsToUMC(decPar, &umcVideoParams);
+    umcVideoParams.numThreads = m_vPar.mfx.NumThread;
+    return MFX_ERR_NONE;
+}
+
+mfxStatus VideoDECODEMJPEGBase_SW::Reset(mfxVideoParam *par)
+{
+    m_tasksCount = 0;
+    pLastTask = NULL;
+    {
+        std::lock_guard<std::mutex> guard(m_guard);
+        while(!m_freeTasks.empty())
+        {
+            delete m_freeTasks.front();
+            m_freeTasks.pop();
+        }
+    }
+
+    memset(&m_stat, 0, sizeof(mfxDecodeStat));
+    m_vPar = *par;
+
+    MFX_CHECK(m_FrameAllocator->Reset() == UMC::UMC_OK, MFX_ERR_MEMORY_ALLOC);
```

Comment:
Please don't put functions call to macro. This one would have additional assert at debug build, which will actually cause first argument to appear twice in macro expansion, i.e. Reset will be called twice. I suggest to save result to variable and then check it